ssz_serialization/codec.nim

```diff
@@ -277,7 +279,89 @@ proc readSszValue*[T](input: openArray[byte],
     if resultBytesCount == maxExpectedSize:
       checkForForbiddenBits(T, input, val.maxLen + 1)
 
-  elif val is HashList | HashArray:
+  elif val is HashList:
+    type E = ElemType(type val)
+
+    when isFixedSize(E):
+      const elemSize = fixedPortionSize(E)
+      when elemSize > 1:
+        if input.len mod elemSize != 0:
+          var ex = new SszSizeMismatchError
+          ex.deserializedType = cstring typetraits.name(T)
+          ex.actualSszSize = input.len
+          ex.elementSize = elemSize
+          raise ex
+
+      val.data.setOutputSize input.len div elemSize
+      when supportsBulkCopy(type val.data[0]):
+        if val.data.len > 0:
+          copyMem addr val.data[0], unsafeAddr input[0], input.len
+
+        # There's no selective invalidation here, because it would require a
+        # potential performance tradeoff, either interfering with bulk copy,
+        # or involving more verification of changed hash entries.
+        val.resetCache()
+      else:
+        var prevValue: E
+        val.resizeHashes()
+
+        for i in 0 ..< val.len:
+          let offset = i * elemSize
+          assign(prevValue, val.data[i])
```

Comment:
since we're optimizing for the non-changing case here, it's better to read into the temporary, compare and then maybe assign it which at least sometimes avoids an assignment

Reply:
That effectively overrides FOOBAR.

I tried this:
```diff
@@ -302,15 +302,15 @@ proc readSszValue*[T](input: openArray[byte],
         # or involving more verification of changed hash entries.
         val.resetCache()
       else:
-        var prevValue: E
+        var readValue: E
         val.resizeHashes()

         for i in 0 ..< val.len:
           let offset = i * elemSize
-          assign(prevValue, val.data[i])
           readSszValue(
-            input.toOpenArray(offset, offset + elemSize - 1), val.data[i])
-          if prevValue != val.data[i]:
+            input.toOpenArray(offset, offset + elemSize - 1), readValue)
+          if readValue != val.data[i]:
+            assign(val.data[i], readValue)
             val.clearCaches(i)
```
But because of the immutable/mutable split (and in general, therefore, nim-ssz-serialization's correctness as a library), it's effectively required here to `readSszValue` directly into the output array, so that it skips updating `dontSerialize` fields entirely, rather than (re)setting them to `default(ValidatorPubKey)` in this case, because that's what `readValue` will have for the immutable validators.

It might be possible to do this optimization this in special cases where it's safe to completely overwrite each time (and/or it's rare enough to run `readSszValue` twice, or specialize `=`/`assign()` or `==` operations to skip fields along the lines of `dontSerialize`) but those are all bigger, riskier changes out of scope of this PR.

Reply:
One way of avoiding the assignment operator I considered, which would be compatible with this and not have these issues, is if one trusts basically just the SHA digests it used by Ethereum, and then one can do a hash comparison.

That might not be faster though, and has its own (small) risks, so, similarly, chose not to pursue that for this PR. If it's enough of a performance win, and that `prevCopy` `assign` proves a bottleneck, then can revisit this.

Reply:
interesting and good point - hm, I kind of agree that solving this properly involves a new set of functions for assignment and comparison that take into account only merkleized fields - I'm not a believer in computing the sha256 for this purpose (it would be unnecessarily slow) - what we would need is something along the lines of an override akin to readSszBytes with this specific purpose ("needsUpdateFromSsszBytes" or whatever) that would do the comparison in the most efficient way for the given type, and then we can introduce special cases one by one - maybe worth exploring, maybe not - unsure.